src/librepgp/repgp.c

```diff
@@ -0,0 +1,369 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <string.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <assert.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+
+#include <rnp/rnp_def.h>
+#include <rnp/rnp.h>
+#include <rnp/rnpcfg.h>
+#include <repgp/repgp.h>
+#include <rekey/rnp_key_store.h>
+
+#include "internal_types.h"
+#include "packet-print.h"
+#include "memory.h"
+#include "utils.h"
+#include "crypto.h"
+#include "reader.h"
+#include "validate.h"
+
+repgp_handle_t
+create_filepath_handle(const char *filename)
+{
+    if (filename == NULL) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    struct repgp_handle *s = calloc(sizeof(struct repgp_handle), 1);
+    if (!s) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    s->filepath = strndup(filename, strlen(filename));
+    s->type = REPGP_HANDLE_FILE;
+    return s;
+}
+
+repgp_handle_t
+create_buffer_handle(const size_t buffer_size)
+{
+    struct repgp_handle *s = calloc(sizeof(struct repgp_handle), 1);
+    if (!s) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    s->buffer.data = (unsigned char *) malloc(buffer_size);
+    if (!s->buffer.data) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    s->buffer.size = buffer_size;
+    s->type = REPGP_HANDLE_BUFFER;
+    return s;
+}
+
+repgp_handle_t
+create_data_handle(const uint8_t *data, size_t data_len)
+{
+    struct repgp_handle *s = calloc(sizeof(struct repgp_handle), 1);
+    if (!s) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    s->buffer.data = (unsigned char *) malloc(data_len);
+    if (!s->buffer.data) {
+        return REPGP_HANDLE_NULL;
+    }
+    memcpy(s->buffer.data, data, data_len);
+
+    s->buffer.size = data_len;
+    s->type = REPGP_HANDLE_BUFFER;
+    return s;
+}
+
+/* Reads into memory everything from stdin */
+repgp_handle_t
+create_stdin_handle(void)
+{
+    char     buf[BUFSIZ * 8];
+    uint8_t *data = NULL;
+    size_t   size = 0;
+    size_t   n;
+
+    struct repgp_handle *s = calloc(sizeof(struct repgp_handle), 1);
+    if (!s) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    /* Read in everything and keeps it in memory.
+     * For stdin it kind of makes sense as no one
+     * should provide a lot of data on stdin.
+     */
+    while ((n = read(STDIN_FILENO, buf, sizeof(buf))) > 0) {
+        /* round up the allocation */
+        size_t   newsize = size + ((n / BUFSIZ) + 1) * BUFSIZ;
+        uint8_t *loc = realloc(data, newsize);
+        if (loc == NULL) {
+            RNP_LOG("Short read");
+            free(data);
+            return REPGP_HANDLE_NULL;
+        }
+        data = loc;
+        memcpy(data + size, buf, n);
+        size += n;
+    }
+
+    s->type = REPGP_HANDLE_BUFFER;
+    s->buffer.size = size;
+    s->buffer.data = data;
+    return s;
```

Comment:
You haven't check that `read` may return error here

Reply:
That's old ``stdin_to_mem`` function which also does not check it. Indeed, to be fixed.